sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala

```diff
@@ -138,6 +138,12 @@ trait DataSourceV2ScanExecBase extends LeafExecNode {
     }
   }
 
+  override def outputOrdering: Seq[SortOrder] = scan match {
+    case s: SupportsReportOrdering if this.logicalLink.isDefined =>
```

Comment:
This is not correct since Spark can combine input partitions (see `groupPartitions`).

Reply:
Combining the partitions through `groupPartitions` does not preserve order (when concatenated), that is right. It could be preserved by merging the partitions, though.

The `outputOrdering` of the data source is as stated. Whoever calls into `groupPartitions` should consider that order is not preserved (depending on how the `Seq[InputPartition]` is used) if existing order is desired.

Reply:
> Whoever calls into groupPartitions should consider ...

Yes, although `groupPartitions` is called by Spark right now so we should update accordingly in this PR too. Otherwise, after this PR the `outputOrdering` could be incorrect if grouping happen.

Reply:
Alright, when partition grouping happens and multiple partitions are grouped, ordering cannot be preserved.

I have modified `outputOrdering` not to return the ordering in that case. Making grouped partitions preserve order is out-of-scope of this PR.